Allow user to ramp workers up/down each iteration

jberkhahn opened the pull request:
modified ExecuteConcurrently() to limit workers throughput based on a supplied pattern in the form of a comma separated list, i.e. -pattern=1,2,3. Default behavior in the case of a missing/malformed pattern is to simply run each iteration with all the workers.

julz commented:
Why a pattern and not just from .. to?

jberkhahn commented:
The story said that linear ramp/up down was fine, but that I should maybe consider "periodic increase/decrease......peaks/troughs". I figured it would be easiest to just let the user specify an array as that would allow for arbitrary patterns.

julz commented:
Let's keep it simple.

julz commented on experiment/runner_test.go:
```diff
@@ -241,7 +266,7 @@ var _ = Describe("ExperimentConfiguration and Sampler", func() {
 		})
 	})
-
+/**
```
If you want to pend a test, use PIt, don't comment it out, if a test is dead code that we don't plan to resurrect then remove it (that's what git is for)

julz commented on server/server.go:
```diff
@@ -128,10 +128,17 @@ func (ctx *context) handlePush(w http.ResponseWriter, r *http.Request) (interfac
+	}
+
+	//ToDo (simon): interval and stop is 0, repeating at interval is not yet exposed in Web UI
+	//ToDo (jonathan): pattern is set to empty string, worker rampup not yet exposed in Web UI
```
instead of a todo, let's create a pending test (with no implementation, but it's better than a dead comment)

julz commented on experiment/runner_test.go:
```diff
@@ -131,6 +131,31 @@ var _ = Describe("ExperimentConfiguration and Sampler", func() {
+			Ω(execExp.pattern).Should(BeNil())
+		})
+	
+		It("Creates and populates a nil channel when the pattern has the incorrect number of iterations", func() {
```
this is just going to confuse users. let's go with e.g. `-concurrency=1..5` and just create the number of workers by interpolating over the number of experiments to get the effective "pattern". I much prefer that to the code silently just not ramping if the number of elements in pattern isn't the same as the number of iterations.

julz commented on cmdline/cmdline.go:
```diff
@@ -31,6 +32,7 @@ func InitCommandLineFlags(config config.Config) {
 	config.StringVar(&params.workload, "workload", "gcf:push", "a comma-separated list of operations a user should issue (use -list-workloads to see available workload options)")
 	config.IntVar(&params.interval, "interval", 0, "repeat a workload at n second interval, to be used with -stop")
 	config.IntVar(&params.stop, "stop", 0, "stop a repeating interval after n second, to be used with -interval")
+	config.StringVar(&params.pattern, "pattern", "", "a comma-separated list of how many workers each iteration should use")
```
how about if we just allow the existing concurrency param to be a string with a format like e.g. `-concurrency-1..5`? (meaning start at 1 worker at iteration 1, move to 5 workers by iteration N). Then you can specify either "2" or "2..2" (which would be equivalent) to get a constant number of workers. I think that's simpler and clearer.

julz commented on benchmarker/benchmarker.go:
```diff
@@ -89,12 +89,14 @@ func Execute(tasks <-chan func()) {
+func ExecuteConcurrently(totalWorkers int, workerPattern <-chan int,  tasks <-chan func()) {
+ 	var wg sync.WaitGroup
+	workers := totalWorkers
+	if workerPattern != nil { workers = <-workerPattern }
```
lets not have nils, let's make the same implementation work for both cases (a non-ramping concurrency is a special case of one that changes over time)

julz replied:
unless Im misreading this (which is possible), you're pulling the number of workers from workerPattern outside the loop. which means it can only be happening once. which means the number can't be changing. which means there's no ramping going on in this function?

jberkhahn commented:
Ok, revised version is up. -concurrency can now be either a single number, or in the form '1..5', where the first number is the # of workers to start, and the second # is the number of workers to ramp up to. Currently only supports ramping up. Added a new command line option -ramptime, which is a single integer that represents how much time, in seconds, between adding new workers.

julz commented on benchmarker/benchmarker_test.go:
```diff
@@ -114,12 +123,26 @@ var _ = Describe("Benchmarker", func() {
 				Ω(result.Seconds()).Should(BeNumerically("~", 2, 1))
 			})
+
+			It("Ramps up the workers", func() {
```
this doesn't really test ramping up

bravera replied:
I am out of the office until 22/04/2014.

For urgent issues regarding the IBM CloudFoundry Community Development team
please contact Alex Tarpinian/Austin/IBM.

For other issues please contact my manager, David Watson/UK/IBM.

Note: This is an automated response to your message  "Re: [pat] Allow user
to ramp workers up/down each iteration (#69)" sent on 11/04/2014 22:37:01.

This is the only notification you will receive while this person is away.

julz commented on benchmarker/benchmarker.go:
```diff
@@ -89,16 +95,34 @@ func Execute(tasks <-chan func()) {
+		tick := time.NewTicker(info.RampTime)
+		for _ = range tick.C {
+			wg.Add(1)
+			go func(t <-chan func()) {			
```
would be good to get rid of the duplication here

bravera replied:
I am out of the office until 22/04/2014.

For urgent issues regarding the IBM CloudFoundry Community Development team
please contact Alex Tarpinian/Austin/IBM.

For other issues please contact my manager, David Watson/UK/IBM.

Note: This is an automated response to your message  "Re: [pat] Allow user
to ramp workers up/down each iteration (#69)" sent on 11/04/2014 22:37:04.

This is the only notification you will receive while this person is away.

julz commented on benchmarker/benchmarker.go:
```diff
@@ -89,16 +95,34 @@ func Execute(tasks <-chan func()) {
 			}
 		}(tasks)
 	}
+	if info.StartingWorkers != info.TotalWorkers && info.RampTime > time.Duration(0) {
```
where's the test for this?

bravera replied:
I am out of the office until 22/04/2014.

For urgent issues regarding the IBM CloudFoundry Community Development team
please contact Alex Tarpinian/Austin/IBM.

For other issues please contact my manager, David Watson/UK/IBM.

Note: This is an automated response to your message  "Re: [pat] Allow user
to ramp workers up/down each iteration (#69)" sent on 11/04/2014 22:37:08.

This is the only notification you will receive while this person is away.

julz commented on benchmarker/benchmarker.go:
```diff
@@ -89,16 +95,34 @@ func Execute(tasks <-chan func()) {
+	for i := 0; i<info.StartingWorkers; i++ {
 		wg.Add(1)
-		go func(t <-chan func()) {
+		go func(t <-chan func()) {			
```
make sure to run `go fmt` to get rid of extra whitespace

bravera replied:
I am out of the office until 22/04/2014.

For urgent issues regarding the IBM CloudFoundry Community Development team
please contact Alex Tarpinian/Austin/IBM.

For other issues please contact my manager, David Watson/UK/IBM.

Note: This is an automated response to your message  "Re: [pat] Allow user
to ramp workers up/down each iteration (#69)" sent on 11/04/2014 22:37:09.

This is the only notification you will receive while this person is away.

julz commented on benchmarker/benchmarker.go:
```diff
@@ -16,6 +16,12 @@ type IterationResult struct {
 	Error    *EncodableError
 }
 
+type ConcurrencyInfo struct {
```
is the struct really helping here? at the very least don't name it "Info" (it's nothing to do with info) and don't needlessly export it outside the module. It might make sense instead of the struct to pass a `chan(workers int)` which would send a number of workers whenever the numbers of workers should change, you could then have a function to create such a channel like `RampingFrom(from int, to int, step int) chan(int)` and then ExecuteConcurrently could just pull from that channel and ramp to the workers requested, that would be a useful abstraction. Or just pass the parameters as parameters, but I don't think a data-only parameter object struct is helpful here.

bravera replied:
I am out of the office until 22/04/2014.

For urgent issues regarding the IBM CloudFoundry Community Development team
please contact Alex Tarpinian/Austin/IBM.

For other issues please contact my manager, David Watson/UK/IBM.

Note: This is an automated response to your message  "Re: [pat] Allow user
to ramp workers up/down each iteration (#69)" sent on 11/04/2014 22:37:54.

This is the only notification you will receive while this person is away.

julz commented on server/server.go:
```diff
@@ -128,10 +128,16 @@ func (ctx *context) handlePush(w http.ResponseWriter, r *http.Request) (interfac
 		NewRunnableExperiment(
 			NewExperimentConfiguration(
-				pushes, concurrency, interval, stop, ctx.worker, workload)))
+				pushes, concurrency, interval, stop, ctx.worker, workload, rampTime)))
```
would be good if the order of the struct fields was [params], [worker], [workload]

bravera replied:
I am out of the office until 22/04/2014.

For urgent issues regarding the IBM CloudFoundry Community Development team
please contact Alex Tarpinian/Austin/IBM.

For other issues please contact my manager, David Watson/UK/IBM.

Note: This is an automated response to your message  "Re: [pat] Allow user
to ramp workers up/down each iteration (#69)" sent on 11/04/2014 22:37:58.

This is the only notification you will receive while this person is away.

julz commented on experiment/runner_test.go:
```diff
@@ -131,6 +131,23 @@ var _ = Describe("ExperimentConfiguration and Sampler", func() {
+			executor := runExp.executerFactory(nil, nil, nil, nil)
+			execExp := executor.(*ExecutableExperiment)
+	
+			Ω(execExp.info.StartingWorkers).Should(Equal(3))
```
experiment.info.StartingWorkers.. why does experiment have an 'info' field that's actually just the concurrency configuration?

bravera replied:
I am out of the office until 22/04/2014.

For urgent issues regarding the IBM CloudFoundry Community Development team
please contact Alex Tarpinian/Austin/IBM.

For other issues please contact my manager, David Watson/UK/IBM.

Note: This is an automated response to your message  "Re: [pat] Allow user
to ramp workers up/down each iteration (#69)" sent on 11/04/2014 22:39:20.

This is the only notification you will receive while this person is away.

julz commented on experiment/runner.go:
```diff
@@ -91,7 +95,12 @@ func NewRunnableExperiment(config ExperimentConfiguration) *RunnableExperiment {
 
 func (c ExperimentConfiguration) newExecutableExperiment(iterationResults chan IterationResult, errors chan error, workers chan int, quit chan bool) Executable {
-	return &ExecutableExperiment{c, iterationResults, workers, quit}
+	workersToUse := strings.Split(c.Concurrency, ".")
```
lets do the parsing in cmdline rather than here, the web UI probably wants to specify starting and total separately.. the use of '..' in the cmdline flags is nothing this module should need to know or care about

bravera replied:
I am out of the office until 22/04/2014.

For urgent issues regarding the IBM CloudFoundry Community Development team
please contact Alex Tarpinian/Austin/IBM.

For other issues please contact my manager, David Watson/UK/IBM.

Note: This is an automated response to your message  "Re: [pat] Allow user
to ramp workers up/down each iteration (#69)" sent on 11/04/2014 22:41:18.

This is the only notification you will receive while this person is away.

julz commented:
this is mostly good I think, regardless the number of mostly minor comments I just made

bravera commented:
I am out of the office until 22/04/2014.

For urgent issues regarding the IBM CloudFoundry Community Development team
please contact Alex Tarpinian/Austin/IBM.

For other issues please contact my manager, David Watson/UK/IBM.

Note: This is an automated response to your message  "Re: [pat] Allow user
to ramp workers up/down each iteration (#69)" sent on 11/04/2014 22:05:13.

This is the only notification you will receive while this person is away.

jberkhahn commented:
OK, new version is cleaned up and posted.

jberkhahn commented:
Ok, new version is up. Note that this requires an update of the version of Ginkgo we use, which I've updated in the Cartridge.lock. Updating Ginkgo didn't actually break anything, that error turned out to be due to a related but separate issue with the dependencies on my local dev environment.

julz commented on ui/js/tests.js:
```diff
@@ -25,7 +25,7 @@ describe("The view", function() {
 
   describe("showThroughput()", function() {
```
did this file actually change? I can't tell because of all the whitespace diffs

julz commented on server/server.go:
```diff
@@ -129,10 +137,11 @@ func (ctx *context) handlePush(w http.ResponseWriter, r *http.Request) (interfac
 		workload = "push"
 	}
 
+	//ToDo (simon): interval and stop is 0, repeating at interval is not yet exposed in Web UI
```
please can we kill this comment? todos go in tracker, not code; also it's no longer true - interval and stop are parsed above

julz commented on cmdline/cmdline.go:
```diff
@@ -41,19 +46,23 @@ func RunCommandLine() error {
 		return validateParameters(worker, func() error {
 			return store.WithStore(func(store Store) error {
 
+				rawConcurrency := strings.Split(params.concurrency, "..")
```
better to use `SplitN` here, also probably should pull these three lines out in to its own function

julz commented on cmdline/cmdline.go:
```diff
@@ -25,7 +29,8 @@ var params = struct {
 	config.IntVar(&params.iterations, "iterations", 1, "number of pushes to attempt")
-	config.IntVar(&params.concurrency, "concurrency", 1, "max number of pushes to attempt in parallel")
+	config.StringVar(&params.concurrency, "concurrency", "1", "number of workers to execute the workload in parallel")
+	config.IntVar(&params.rampStepTime, "rampsteptime", 60, "seconds between adding additonal workers when -concurrency specifies more than one value")
```
`concurrency:timeBetweenSteps` ? I still think this is probably clearer specified as sthg more like `concurrency:endTime`, `concurrency:endWorkers`, `concurrency:startWorkers`, but timeBetweenSteps is at least fairly clear to someone who has no idea what "ramp" means (and why should they, since we never use the word 'ramp' anywhere else)

julz commented on cmdline/cmdline.go:
```diff
@@ -15,7 +18,8 @@ import (
 	listWorkloads bool
-	concurrency   int
+	concurrency   string
+	rampStepTime      int
```
go fmt

julz commented on benchmarker/benchmarker_test.go:
```diff
@@ -101,24 +101,55 @@ var _ = Describe("Benchmarker", func() {
+						loopCounter <- i
+						tasks <- func(n int) {
+							defer GinkgoRecover()
+							Ω(executionCounter).Should(Equal(<-loopCounter))
```
we can probably make it easier to follow what this code is doing by more carefully choosing the variable names. For example `Ω(orderWasExecuted).Should(Equal(orderWasAddedToQueue))`

julz commented on benchmarker/benchmarker_test.go:
```diff
@@ -101,24 +101,55 @@ var _ = Describe("Benchmarker", func() {
+				go func() {
+					loopCounter := make(chan int, 1)
+					defer close(tasks)
+					defer close(loopCounter)
```
you don't need to close channels that nothing is ranging over (closing is just a signal that anything that's ranging, or in other ways checking the close state of the channel, can use to know there aren't going to be more values)

julz commented on benchmarker/benchmarker.go:
```diff
@@ -89,16 +89,19 @@ func Execute(tasks <-chan func(int)) {
-		}(tasks, i)
+
+	for increment := range schedule {
+		for i:=0; i<increment; i++ {
```
I still don't think this line is tested anywhere; you only test the case when a 1 is sent on the channel, never the case where the channel asks for a step greater than 1. Either add a test, or lose this line

jberkhahn commented:
Ok, added an additional test case to test for schedule events >1, renamed a bunch of the variables. I also moved the parsing for the step time into it's own method, as in the future we're probably going to change that around for different ramp patterns.

julz commented on cmdline/cmdline.go:
```diff
@@ -13,19 +16,21 @@ import (
 	config.IntVar(&params.iterations, "iterations", 1, "number of pushes to attempt")
-	config.IntVar(&params.concurrency, "concurrency", 1, "max number of pushes to attempt in parallel")
+	config.StringVar(&params.concurrency, "concurrency", "1", "number of workers to execute the workload in parallel")
+	config.IntVar(&params.concurrencyStepTime, "concurrency_timeBetweenSteps", 60, "seconds between adding additonal workers when ramping works up")
```
for consistency with things like `rest:password` and `rest:username`, let's use `concurrency:timeBetweenSteps`?

julz commented on experiment/runner.go:
```diff
@@ -46,7 +47,8 @@ type Experiment interface {
 	Iterations  int
-	Concurrency int
+	Concurrency []int
+	ConcurrencyStepTime    time.Duration
```
go fmt

julz commented:
couple inline comments; could you try to set up whatever editor you're using to do a 'go fmt' when you save, or set up a git hook or something to 'go fmt' before commit (if you're using vim then [go-vim](https://github.com/fatih/vim-go) looks like a good way of getting that working). Other than that looks good.

jberkhahn commented:
New version up. Left nested-yaml parsing for another story, as it deals with a bunch of stuff regarding parsing params that are a bit outside the context of this story.

julz commented:
ok, looks good, seems a bit weird that the commit is in simon's name; do you want to do `git commit --amend --reset-author` before I merge?

jberkhahn committed:
modified ExecuteConcurrently to ramp workers up based on a provided
event queue

jberkhahn commented:
Ok, amended the author.
